### DISCUSSION TITLE
Proposal: Button Entity

### DISCUSSION THREAD

**Discussion by frenck** [home-assistant/architecture#643] [Entity Models] [answered]
## Context

This is inspired by a switch on my desk (that needs to be installed) and by what was found during the development of some things with Tuya the past few days (Tuya has a LOT of devices).

**Home Assistant is lacking a concept for user input: Button. Also known as momentary switches, push-buttons, or stateless switches.**

Such a button can be useful for "triggering" virtually anything, that doesn't maintain its state (stateless). Examples of such actions include:

- Upgrade firmware
- Reboot/Restart a device or service
- Brew a cup of coffee
- Reset something (like a counter, filter usage, ...)
- Confirming an action or event (e.g., Vacuum cleaner bin is emptied)

## Current situation

As we don't have this concept right now, we often implement one of the following solutions:

- **We use a switch/toggle.**
  ESPHome is a good example of this, it uses a switch as a restart/reboot entity. Once you toggle it... *poof* device unavailable. After it restarted, the switch is off again. You barely see the switch turn on actually.

    ![image](https://user-images.githubusercontent.com/195327/137345292-4117eeeb-1a2c-4c54-af8c-0925d868313d.png)

- **We implement a service / we use a service call.**
  This has been the most common method lately. Implementing services for triggering stateless actions.
  While service are cool, they are slightly more complex/less user friendly (compared to the switch example of ESPHome above).
  This is mainly caused by each service being unique for that integration and therefore there is a lot of services out there.

We leave it up to the user to make something of it. Funny enough, we do support buttons in our Lovelace UI.

![image](https://user-images.githubusercontent.com/195327/137345398-4aeee84c-db7e-4212-b58e-ca15e612570b.png)

These buttons can be used as a push-button to trigger something and can be stateless, but to require such a thing to be set up by the user. E.g., by defining a custom action to... a service call 😬 

### TL;DR:

So, while possible to create some kind of momentary-push-button-trigger-thingy, it is up to the user to create something like that. Meanwhile, we use custom services or misuse switches to work around the fact we lack a trigger input from Home Assistant.

## Proposal

I like to propose a new entity: **Button** ▶️ 

- The button entity can be pressed to fire an even towards a device or service but remains stateless from a device or service perspective.
- In Home Assistant we use the unused state for storing a date timestamp to indicate when the button was last pressed. This is automatically and only is the button is pressed in the UI or via service call from Home Assistant.
- Buttons have 1 service: `button.press` to press the button via e.g., automations or scripts.
- Can be automated on, it is able to trigger automatoins on the fact that a button has been pressed.
- Can be really helpful in providing actions button to complement the [config entities merged recently]( https://github.com/home-assistant/core/pull/57528)

From the user interface the button could be similar to e.g., activating a scene. Example mockup:

![image](https://user-images.githubusercontent.com/195327/137347448-0f33c668-d7a2-4073-a960-9fe4d85e31d7.png)

IMHO, it would be wise to replace the "press" text with an ▶️  sort of icon. Or maybe make the text used as the action configurable in some way. 

## Extending on Button

If the button lands in the Home Assistant eco-system, we can start building on it.
For example:

- ~~Introduce a `input_button`~~ -> Script.
- Introduce a button template, that allows for defining their own action on a button entity.
- Allow ESPHome users to define buttons (instead of the current service method that creates services for each device).
- .... bunch of more integrations to be adjusted.

## What a button is not

- It is not a replacement for a binary sensor, as a button can be actioned (input), while a binary sensor purely is a state (output).
- It is not a replacement for events created by physical switches (for which we misused binary sensors back in the day). As this entity is actionable virtually (pressable), which would not happen in the physical world. So, while events are used for Device -> Home Assistant; buttons are used for Home Assistant -> Device.

## Consequences

It needs to be implemented, documented and such.

https://github.com/home-assistant/core/pull/57642 👀

**Comment by dgomes** [comment]
How does it relate to https://developers.home-assistant.io/docs/core/entity/remote/ ?

**Reply by frenck** [reply]
Remote is resembling a "Remote" device with multiple buttons. It aimed towards a television remote (or similar). It can learn commands and provides default actions to turn things on/off (and has a state). It can be used to learn activities and later replay the things it learned (e.g., codes, sequences) again or provide multiple activities. The current activity is also a state of the remote.

A button does 1 thing. Fire something.

**Reply by dgomes** [reply]
But couldn't such "Remote" be implemented through various "Buttons" ? 

BTW: some remotes don't have a state (broad link), some don't learn (its a optional feature) and it does not have a state just like the proposed button

**Reply by frenck** [reply]
No, as button doesn't reflect a state.

**Reply by dgomes** [reply]
Remote "buttons" (aka commands) don't have a state

I just think these concepts overlap, but I like this simpler approach

**Reply by hunterjm** [reply]
I agree with @dgomes here. I don't want to hijack this architecture discussion to talk about remote, but it is pretty similar to the concept of `command` in the remote entity. While the remote can and does hold state in some integrations (xbox, appletv) it does not in others (broadlink).

It has always bugged me that most of the time, valid "commands" are known by the integration but not able to be represented in any way in the UI. We have resorted to [listing](https://www.home-assistant.io/integrations/apple_tv/#remote) [them](https://www.home-assistant.io/integrations/xbox/#remote) in the documentation or pulling them from the [device](https://github.com/home-assistant/core/blob/dev/homeassistant/components/broadlink/remote.py#L134-L171) not to display them to the user, but to ensure that a valid command is sent.

I'm all on board with this architecture discussion, but when approved and implemented we should think about how to integrate buttons with remotes. Personally, Xbox has a `media_player` entity to handle turning it on and off. The only reason a `remote` exists is to send commands to the console and it would probably be more user friendly to implement those as `button` entities if this proposal is accepted unless `remote` is also updated to implement commands as buttons.

**Comment by TomBrien** [comment]
What do you see an automation triggered by a `button` looking like? Do we just trigger from the event fired by the button or do we introduce a new `button_press` trigger?

**Reply by frenck** [reply]
We can use the state trigger already, but a dedicated trigger for the button can be done if that makes things clearer.

```yaml
trigger:
   - platform: state
     entity_id: button.reboot_thingy
```

The above-listed/linked PR also contains support for device triggers/actions.

**Comment by thecode** [comment]
Hi @frenck this is a great addition even without any modifications to your suggestion above we should go with it.
Other enhancements I suggest considering (which are more frontend related, but backend may need to support:
- Consider supporting different text for the frontend action, while `PRESS` fits most actions, there are places where `OPEN`, `START`, `GO`, `REBOOT`, etc.. may be a better fit
- Support for different frontend display types

As this discussion is mainly for the entity, the only thing need to be considered is if the above need to be stored in the entity as properties.

**Reply by frenck** [reply]
> Consider supporting different text for the frontend action, while PRESS fits most actions, there are places where OPEN, START, GO, REBOOT, etc.. may be a better fit

Hmmm device classes?

> Support for different frontend display types

That should really be a frontend concern (just like you can change the display of an entity now as well).

**Reply by thecode** [reply]
Device classes sounds perfect and I agree about the frontend, just raised it because a follow up frontend PR will probably happen soon

**Comment by jjlawren** [comment]
Does it make sense to allow context for the button press itself, such as long hold, # of taps, etc? This would be dependent on UI implementation (and likely processed client-side), but it could add some interesting usage options.

**Reply by frenck** [reply]
I really and strongly believe we should keep that out of scope. This is not a frontend-only thing.
Also, this is probably better solved with a card capable of accepting multiple button entities in that case.

That said, if there is a general recurring need for such a thing, we can always add stuff later. We should start minimal :)

**Reply by gonzotek** [reply]
I'd love to see a Press event and a Release event, which would then allow for determining if there was a long press or multi-taps.  Not sure if that falls under the entity definition or another area of HA, but it would have been useful to me in multiple situations over the years.

**Reply by frenck** [reply]
That is not a backend concern :)

**Reply by farmio** [reply]
The option to define different "pressed" and "released" actions would be very handy indeed. I don't get why this wouldn't be a backend concern as there would have to be a method to override by an integration.
With that a button could be used e.g. to move something or activate a door buzzer for as long as the button is held.

From a service perspective this could just default to fire "pressed" and "released" immediately after another (when not used one of them would just be a no-op). Optionally a timespan could be given. Consecutively incoming service calls can reset this time.

Admittedly this would cause this button to have a state "pressed" but this behaviour can not really be done with switch so I think the button entity would be right for it.

**Comment by tetienne** [comment]
It seems what you describe can be achieve by services, but at least a button will be listed in the device dedicated page.

**Reply by frenck** [reply]
Not only that, it can be added to the UI, e.g., using an entities card. Additionally, you can automate the fact it has been pressed/called.

**Reply by raman325** [reply]
just to add to that, there are likely a fair number of services that would be migrated over to button entities.

Also, I use this custom card to emulate this (call a service on click) but I agree that having it be native would be a better experience: https://github.com/custom-cards/button-card

**Comment by balloob** [comment]
I like the idea. Here are a few thoughts:

1. This is about triggering an action in another device. So I am not sure if we should fire an event in HA too? Especially because we can't guarantee to also fire that event when pressed outside HA
2. Keeping track of "last pressed" for the UI is easy to out of sync because if ie. the reboot button is pressed outside HA we can't update the last changed in HA. 
3. We wouldn't need input_button because that's just a script. Maybe a template button if we want to allow creating things under that domain.

**Reply by frenck** [reply]
> This is about triggering an action in another device. So I am not sure if we should fire an event in HA too? Especially because we can't guarantee to also fire that event when pressed outside HA

I share this thought, I've put this in place for creating a consistent experience across the board of entities. For example, if I now change the state of the light in the UI, I can automate that event too.

We have more entities in HA without an actual state (zoning...), however, I think that that should be avoided and instead van be made useful. (Zone should have a person count as state IMHO 😉 )

It is up to the integration implementation to update the button state (in the current implementation), which allows for the integration to fire if it can guarantee it (or be optimistic).

> Keeping track of "last pressed" for the UI is easy to out of sync because if ie. the reboot button is pressed outside HA we can't update the last change in HA.

This is up to the implementation of the integration. The same can be said about toggling a bulb twice. If the device has an event, log, historical state, or maybe even timestamps of occurrences available, it could sync up.

> We wouldn't need input_button because that's just a script.

Agreed.

**Reply by farmio** [reply]
> It is up to the integration implementation to update the button state (in the current implementation), which allows for the integration to fire if it can guarantee it (or be optimistic).

This is not possible in the current implementation as `__last_pressed` is private. 
For integrations having physical devices represented by a button entity (like to reset a timer for example) it would be nice if `async_write_ha_state` (or a new helper method) could trigger a state_changed event so users automating on the button entity can use it to automate on internal and external events.

**Comment by Misiu** [comment]
I know that button entity should be stateless, but what about disabling the button until the called action is done (and is available to be called)?
For example, when you click the `upgrade firmware` button, the button should be blocked until the firmware is updated and the device reboots, or the `brew a cup of coffee` button should be disabled until the device finishes making the coffee.
I'm not sure if this should be done on the frontend or on the backend, but something like a condition or a cooldown should be possible.

**Reply by Jc2k** [reply]
+1

I think it already has to have some state so that it can be "Unavailable" (e.g like every entity is when it's backing device or API is unavailable). At the simplest we could just use that. Or add some explicit states like "pressable" or "unpressable".

Then any of the logic you described can be deferred to the integration to manage by returning the correct state.

**Reply by frenck** [reply]
The available property already allows for that.

**Comment by frenck** [comment]
Based on the feedback from @balloob  above:

<https://github.com/home-assistant/architecture/discussions/643#discussioncomment-1479754>

I've adjusted the PR and text of this architectural suggestion, I've changed the following:

- The state of the entity contains the timestamp of the last state. However, this only represents the timestamp of when the button was last used within Home Assistant (UI pressed or via serviced call). It cannot be provided by the integration.
- The last pressed timestamp is restored


The reason for this change: 

- We can still trigger/automate on this state change (e.g., do additional actions when the button is pressed from automations).
- We make it clear in the docs what this data represents.
- No integrations specific behavior (except for the single `press` method implementation that defines the action).

Updated implementation is here: <https://github.com/home-assistant/core/pull/57642>

**Comment by dkagedal** [comment]
A "button" is a UI element or a hardware gadget. What you're describing sounds like the backend feature that would be activated by pressing the UI button or the physical button. So is "button" really the right term for this kind entity?

From a backend perspective, it is an entity that has a single operation: `activate` (not `press`). Can we choose a terminology that reflects that?

As it often is, naming is hard, and I have no obvious candidate (and some candidates are already used for other meanings). Some (poor) ideas: activator, activatable, signal, trigger, ...

**Reply by frenck** [reply]
`activate` reflects a state, as something can be activated or deactivated. Therefore, I've avoided that term, as this is meant to be stateless.

Trigger would be more fitting, however, that term is really confusing combined with automations.

**Reply by TomBrien** [reply]
As a user a button is something I press/click/tap depending on if it's physical or virtual. Trigger seems like a compromise but I agree it's confusing if you end up with something like`trigger: trigger`. `pressed`, at least in English, seems a good fit.

**Reply by dkagedal** [reply]
As a user, you are using a UI, and there a "button" is a typical UI element you would interact with, yes. But this discussion is about a kind of backend feature (the entity) that is triggered when you hit that button. I think it would be a mistake to call that backend interface a "button".

Maybe I'm fighting windmills here, but I feel in general that the entities are an API surface, and the UI that interacts with them is a separate thing.

**Reply by frenck** [reply]
> Maybe I'm fighting windmills here, but I feel in general that the entities are an API surface, and the UI that interacts with them is a separate thing.

Agree, however, it is also user-facing, as it defines the name of the entities and services. Since it is user-facing, calling it "trigger" is really confusing as triggers are already a core concept of Home Assistant for automations.

**Comment by mib1185** [comment]
Hi @frenck ,
this is a very nice addition 👍 
I have only one thought if it would be possible to add something like an `need_confirm` (_default `False`_) property, so that user will be forced in the UI to confirm - example: when there are multiple "reset XXX" or "reboot yyy" buttons, a confirmation could be enforced, so that an accidental press can be avoided 🤔  this property should only be respected in frontend

**Reply by Bre77** [reply]
That would entirely get implemented in the front-end, so probably shouldn't be part of the entity.

**Reply by frenck** [reply]
Still, the front end cannot guess if it is needed to confirm or not.

That said, this can be solved in different ways:

- An additional property
- A device class

The latter one, could have the benefit of defining a customized and translatable message.

Anyways, I think that these are a typical things we should leave out of an initial implementation. Once this is in, let's work on extending things.

**Reply by mib1185** [reply]
> Anyways, I think that these are a typical things we should leave out of an initial implementation. Once this is in, let's work on extending things.

I fully agree with that 👍 looking forward to contribute on that 🙂

**Comment by frenck** [comment, marked as answer]
The button entity has been merged in: <https://github.com/home-assistant/core/pull/57642>

Developer documentation is available here:

<https://developers.home-assistant.io/docs/core/entity/button/>
